### ISSUE TITLE
[Design] Layouts and Widgets

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/Orchard2#498]
I have given some thoughts about how to improve content placement, which include widgets and layout, as they are known in O1.

TL;DR;

Nothing much to change in Widgets, reuse Widgets as the fragments for Layouts, build a new Layout editor like CraftCms.

## Current state

Here is my current understanding of these two modules in O1.

### Widgets

Be able to place some content in specific locations across pages. This relies on the notion of _layers_ that represent a predicate. When a layer is active, the widgets associated to it are displayed in their configured zones.

A Widget is a stereotype (a category of content types). It has a title (a section header) and a position. It's instances are content items, which can be drafted, versioned and localized.

### Layout

It's a part that lets an editor define organized content in a grid. Its content is part of the content item, so it can be versioned, drafted and localized as a whole. Each fragment of the grid is called an Element. They are a specific concept (different than field, part or widget).

### Pros

- Widgets are great to place content across different pages. The notion of layer is easy to understand and powerful.
- Layout are very simple to use and understand and fulfill a very important requirement from users.
- The layout storage is optimal as it contains all elements in a single json document.
- The layout part and widgets per page have simplified the widget feature as before we would have had created a layer for each page.

### Cons

- Widgets and Elements seem to have something in common but they are yet different. They should be interchangeable, meaning the same component should be reused for Layout or Widgets.
- There is a special part that allows a user to add widgets to a content item, though each widget is still a distinct content item and it not optimal for performance.
- The Layout editor is not adapted for smaller devices and touch screens.

### Widgets/Layers

- Use Widget content types as the common concept for Layouts and Layer items. A Widget now represent any content fragment that can be placed on a page. We can reuse fields parts and blocks even for a Layout. And we can reuse more complex widgets in layouts too (Tags, projections, ...).
- Rename the Widget module to **Layers** as widgets is now just a _stereotype_.
- The Layers module would behave the same way as Widgets in O1, with independent (versionable, draftable, localizable) content items. The layer would store the content item ids to be able to retrieve them in a single request. The script evaluation would still be done in a scripting language and extensible.
- The _Widget Container_ part storage would contain the content of each widget content item it's made of.
- The list of available zones would be defined in the settings of the content type part definition (per content type).

### Layout

- Rename the Layouts module to [TBD] to prevent the confusion with theme layouts.
- Create a Layout part that can be configured with a selection of Widget types. These types are available in the Layout editor. c.f. CraftCMS video: https://craftcms.com/features/matrix.
- The layout part would contain the content of each widget content item it's made of.
- The allowed positions of a widget in a layout would be defined in the settings of the Layout part, common to all the widgets it can contain.

@DanielStolt @bleroy @sfmskywalker

**Comment by sebastienros** [comment]
Layouts -> Flow ?

**Comment by sfmskywalker** [comment]
Flow, as in the name for the Layouts module?

**Comment by DanielStolt** [comment]
We have had a couple of discussions about this offline, so I would like to paraphrase how I interpret this proposal, expressed a little differently. And at the same time add my personal dreams and desires into the mix. :)

## Widget is the new block

Widgets simply become the new "universal block" for every type of reusable/embeddable content. Elements (which are part of layouts today) would no longer exist at all. The distinction between these two is erased.

Just as before, there can be different types of widgets, some very simple (like HTML widget), some more advanced like projections, and you can also build your own. Just as before, they are content items, draftable, versionable, localizable and so on.

These widgets can be used in three ways:

- Placed on **layers** into global zones defined by the theme
- Placed on **flows** locally within a content item
- Rendered imperatively/programmatically from a Razor view

Widgets can be either **global** or **local**. A global widget can be created once and then reused in all the above places. Local widgets are created and managed completely within a specific layer or flow, and are not visible in any global content list or list of global widgets.

## Layers

As Sebastien wrote, layers would be pretty much what the widgets module is today. They become a place to express that a given widget (global or local) should appear in a specific place under a specific set of circumstances.

The editor for layers would be made more interactive with drag-and-drop support, but it would not concern itself with trying to graphically represent the layout. Just like today, the theme manifest declares a number of zones, and the layer editor shows them top-to-bottom, and allows you to place widgets on the intersections of these layers and the zones.

## Flows

Flows would play the same role that layouts play today in O1 (i.e. building the local content of a content item) but with a different philosophy. Instead of trying to be a "layout" editor and concern itself with WYSIWYG graphical representation of the content and its layout, the flow editor is about building your **content, structure and flow**.

Instead we utilize live preview to provide the WYSIWYG part.

Here you can also place widgets into your flow, both global and local. The toolbox lets you access the global widgets, and also lets you create and embed local widgets of all the available widget types.

The flow editor is still very slick and interactive, just like the layout editor today, but it focuses on structuring and organizing your content, rather than laying it out in two-dimensional space. The graphical depiction provided by the layout editor today is never very close to reality anyway, and it's not even theoretically feasible because what the theme will do at runtime is not known at design time.

The live preview can also provide explicit support for selecting which theme to preview with, in cases where your site has multiple active themes and maybe custom theme selector logic.

## Widgets from Razor views

Finally, it would be cool if you could also render a global widget directly from within a Razor view. Have no idea how hard or easy that might be to accomplish, but it would enable some very appealing scenarios IMO.

**Comment by sebastienros** [comment]
> render a global widget directly from within a Razor view

This is super easy to achieve, like it was done to render the Menu using the menu shape, using the content item id or an alias, also reusing caching options, and selection which display type to use. And it's for any content item, not just Widgets.

**Comment by jtkech** [comment]
Just one idea on the fly because i don't remember the details and maybe too late.

Ok about the `Preview` and unification. So, maybe we could always use the main concept of content items in which, with the flow editor, we define zones in which we add items ... The idea being to also use the flow editor for theme zones.

- So, Maybe a theme could be tied to a content item where, with the flow editor, we could define e.g zones placement properties. Maybe we couldn't create them if still provided by the theme manifest.

- Then, in each zone we could add items with placement properties, layer rules ... Then we could do the same in an individual item.

- I say this because somewhere it's quite the same to add a widget in a theme zone as defining an area in a page content item where we place e.g another widget.

-  And somewhere it's also quite the same to have a widget in a theme zone and with a layer rule as having a page content item in the content zone with the implicit rule related to the current request.

- E.g in O1 with layout we can define in items what is normally done through theme zones. But the need of layout templates which is not so simple e.g when updating a template and then all its related items.

- The problem was because we are doing in an item what is normally done through theme zones, and this because we couldn't use the power of the layout editor for zones of a given theme (not tied to an item).

Best.

**Comment by sebastienros** [comment]
If you mean being able to add a widget to a zone, from a content item editor, we have that in mind too. Sipke made a module for that in O1.

**Comment by jtkech** [comment]
Yes i know so we can do things as normally done with theme zones in the widgets page. But when we want to do the exact same things for multiples items, there is the need of having layout templates.

But layout templates are not simple to implement because e.g when you update a template you need to update all its related items. So the idea to apply the flow editor directly to the theme zones, a theme being tied to a kind of content item.

**Comment by sebastienros** [comment]
skype me, I don't understand

**Comment by jtkech** [comment]
But Sebastien explained to me that also using the flows editor for the placement of theme zones themselves would be too limited comparing to e.g what we can do with razor files.
